### PR TITLE
Complete rework to GL 4 and GLES and synchronize with current rsxs version

### PR DESCRIPTION
**Pull request by AlwinEsch**
This is the complete rework of the addon and now with 22 screensavers. 

The only one who not brought in from http://rss-glx.sourceforge.net/ is the PixelCity, this was to complex and with the Python addon Aerial not direct needed (for me).

All work now with OpenGL 4. But still need fixes and a bit cleanup. Also are GLES changes not done complete.

If this changes are finished is the addon supported on Linux, Mac and Android

### CODE REVIEW

**Comment by AlwinEsch** [comment]
Do not know if it fits, but at least Travis is now running through. :-D

He now builds ImageMagick7 himself to have the code compatible with Mac.

This is reduced to a minimum and has only the following dependencies: "-ltiff -ljpeg -lpng16 -lz -lltdl -lm".

Think so it should also be easier to create on Android.

**Comment by AlwinEsch** [comment]
Now have the missing parts switched to OpenGLES.

At the test here everything went (with slight restrictions).

Only I saw in the render performance that it did not run smoothly (on normal system with Ubuntu 18.10). I do not know if this is related to my system.

**Comment by AlwinEsch** [comment]
From the processing ago, it is now so far final. Has been tested under GL, GLES and with Wayland, ran smoothly everywhere.

Only GLES required some acceptable limitations.

**Comment by AlwinEsch** [comment]
The request under Kodi is really required to have this working!!!

**Comment by AlwinEsch** [comment]
Is updated again, now it is checked if "glm" and "gli" exists in the system, if not it is created by itself. So it should be easy to build for Kodi's build system and an outside user.

Only with ImageMagick I'm not sure.

Would somehow keep the current version (not my problem if Ubuntu is behind). Furthermore, I think it is a must to create this yourself because otherwise too many dependencies are needed (which would never be used in Kodi).

Is there for:
```
ExternalProject_Add(ImageMagick7
                    GIT_REPOSITORY    https://github.com/ImageMagick/ImageMagick
                    GIT_TAG           7.0.8-25
...
```
OK?

**Comment by notspiff** [comment]
No. Everything as depends. Use established mechanisms for this. See my pr, and extend it. Absolutely NO downloading in main build system.

**Comment by AlwinEsch** [comment]
Is updated again. Removed the GLM and GLI self build. Fixed the debian build, what is here more complex and also has needed the pseudo "screensavers.rsxs/screensavers.rsxs/addon.xml.in" to give the jenkins a base to scan for debian source package.

The only part who still as external in ./lib is ImageMagick and becomes normally ignored. Only in case by manual build is there is no ImageMagick present or it becomes forced with `-DOWN_IMAGE_MAGICK=1` on cmake.

**Comment by AlwinEsch** [comment]
I think the changes are now finished :-D

**Comment by AlwinEsch** [comment]
Is this request OK to come in?

**Comment by notspiff** [comment]
let's see ;)

**Comment by AlwinEsch** [comment]
Thanks a lot! :D

File: .gitignore [hunk 1]
```diff
@@ -26,5 +37,4 @@ screensaver.rsxs.*/addon.xml
 # clion
 .idea/
 
-# KDev
-*.kdev4
\ No newline at end of file
+*.patch
```
Comment (Rechi) [inline .gitignore:40]: Ignoring patch files makes workings with depends complicated as you have to use `git add -f`.

File: .travis.yml [hunk 2]
```diff
@@ -1,4 +1,4 @@
-language: cpp
++language: cpp
```
Comment (Rechi) [inline .travis.yml:1]:
```suggestion
language: cpp
```

File: .travis.yml [hunk 3]
```diff
@@ -31,7 +31,8 @@ matrix:
 before_install:
   - if [[ "$TRAVIS_OS_NAME" == "osx" ]]; then which cmake || brew update        ; fi
   - if [[ "$TRAVIS_OS_NAME" == "osx" ]]; then which cmake || brew install cmake ; fi
-  - if [[ "$TRAVIS_OS_NAME" == "osx" ]]; then brew update && brew upgrade cmake || true; fi
+  - if [[ "$TRAVIS_OS_NAME" == "osx" ]]; then brew update && brew upgrade cmake libpng libtiff || true; fi
+  - if [[ "$TRAVIS_OS_NAME" == "osx" ]]; then brew install libtool openjpeg xz ; fi # depends for ImageMagick 7
```
Comment (Rechi) [inline .travis.yml:35]: If those libs are required, they have to be added to depends.

File: CMakeLists.txt [hunk 4]
```diff
@@ -1,192 +1,94 @@
...
 
-list(APPEND CMAKE_MODULE_PATH ${CMAKE_SOURCE_DIR})
+cmake_policy(SET CMP0054 NEW)
+
+option(OWN_IMAGE_MAGICK "Build own reduced ImageMagick" OFF)
 
 find_package(Kodi REQUIRED)
-find_package(OpenGL REQUIRED)
+find_package(ImageMagick COMPONENTS MagickCore)
+find_package(ImageMagick COMPONENTS MagickWand)
```
Comment (Rechi) [inline CMakeLists.txt:14]:
```suggestion
 find_package(ImageMagick REQUIRED COMPONENTS MagickCore MagickWand)
```

File: CMakeLists.txt [hunk 5]
```diff
@@ -1,192 +1,94 @@
...
-endif()
-set(configure_start ${PROJECT_SOURCE_DIR}/${rsxs_dir}/configure)
-if(DEPENDS_PATH)
-  set(PNG_PATH "--with-png=${DEPENDS_PATH}")
-  if(EXISTS ${DEPENDS_PATH}/share/config.site)
-    set(configure_start CONFIG_SITE=${DEPENDS_PATH}/share/config.site ${configure_start})
-  endif()
+find_package(LTDL)
+
+# set(APP_RENDER_SYSTEM "gles") # Leaved here for test purpose only
```
Comment (Rechi) [inline CMakeLists.txt:22]: remove

File: CMakeLists.txt [hunk 6]
```diff
@@ -1,192 +1,94 @@
...
+  add_definitions(${OPENGLES_DEFINITIONS})
 endif()
-if(APPLE)
-  set(EXTRA_FLAGS "CFLAGS=${CMAKE_C_FLAGS} -fPIC -isystem${DEPENDS_PATH}/include -isystem${KODI_INCLUDE_DIR}" "CXXFLAGS=${CMAKE_CXX_FLAGS} -fPIC -isystem${DEPENDS_PATH}/include -isystem${KODI_INCLUDE_DIR}")
+
+include_directories(${KODI_INCLUDE_DIR}/.. # Hack way with "/..", need bigger Kodi cmake rework to match right include ways
+                    ${GLM_INCLUDE_DIR}
+                    ${GLI_INCLUDE_DIR})
+
+if(ImageMagick_FOUND AND NOT ${OWN_IMAGE_MAGICK})
```
Comment (Rechi) [inline CMakeLists.txt:40]: Why is this needed?

File: debian/changelog [hunk 7]
```diff
@@ -1,11 +1,7 @@
-kodi-screensavers-rsxs (1.0.1-1~raring) raring; urgency=low
+kodi-screensavers-rsxs (3.0.0-0~cosmic) cosmic; urgency=low
```
Comment (Rechi) [inline debian/changelog:1]: This file shouldn't be committed, because it is autogenerated.

File: debian/compat [hunk 8]
```diff
@@ -1 +1 @@
-8
+11
```
Comment (Rechi) [inline debian/compat:1]: This breake Ubuntu Xenial (16.04)

File: debian/rules [hunk 9]
```diff
@@ -9,23 +9,44 @@
 # Uncomment this to turn on verbose mode.
 #export DH_VERBOSE=1
 
+DEB_HOST_MULTIARCH ?= $(shell dpkg-architecture -qDEB_HOST_MULTIARCH)
+
 %:
-	dh $@ 
+	@for target in `echo $(shell find debian/*.install.in)`; do \
```
Comment (Rechi) [inline debian/rules:15]: Where the install targets with the asterisks broken, that this is needed?

File: depends/common/gli/gli.txt [hunk 10]
```diff
@@ -0,0 +1 @@
+gli https://github.com/g-truc/gli/archive/dd17acf9cc7fc6e6abe9f9ec69949eeeee1ccd82.tar.gz
```
Comment (Rechi) [inline depends/common/gli/gli.txt:1]: Does the latest release not work?

File: depends/common/glm/glm.txt [hunk 11]
```diff
@@ -0,0 +1 @@
+glm https://github.com/g-truc/glm/archive/47031aa4b7b079716e4cac496819e7f464b6a713.tar.gz
```
Comment (Rechi) [inline depends/common/glm/glm.txt:1]: Does the latest release not work?

File: depends/common/imagemagick/CMakeLists.txt [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+include_directories(${LTDL_INCLUDE_DIRS} ${JPEG_INCLUDE_DIRS} ${PNG_INCLUDE_DIR} ${TIFF_INCLUDE_DIR})
+
+message(STATUS "--------------------------_> ${CMAKE_MODULE_PATH}")
+# Force on travis where only Ubuntu 16.04 Xenial is available the use of gcc
+# on Image Magick, otherwise brings it configure errors!
+if(DEFINED ENV{TRAVIS_COMPILER})
+  if($ENV{TRAVIS_COMPILER} STREQUAL "clang")
+    set(OVERRIDE_COMPILER CC=gcc CXX=g++)
+  endif()
+endif()
```
Comment (Rechi) [inline depends/common/imagemagick/CMakeLists.txt:19]: If building with clang is broken, just remove the whole job from travis.

File: depends/common/jpeg-9b/jpeg-9b.txt [hunk 13]
```diff
@@ -0,0 +1 @@
+jpeg-9b http://mirrors.xbmc.org/build-deps/sources/jpegsrc.v9b.tar.gz
```
Comment (Rechi) [inline depends/common/jpeg-9b/jpeg-9b.txt:1]:
```suggestion
jpeg-9b http://mirrors.kodi.tv/build-deps/sources/jpegsrc.v9b.tar.gz
```

File: depends/common/libtool/libtool.txt [hunk 14]
```diff
@@ -0,0 +1 @@
+libtool http://mirrors.xbmc.org/build-deps/sources/libtool-2.4.6.tar.gz
```
Comment (Rechi) [inline depends/common/libtool/libtool.txt:1]:
```suggestion
libtool http://mirrors.kodi.tv/build-deps/sources/libtool-2.4.6.tar.gz
```

File: depends/common/png/png.txt [hunk 15]
```diff
@@ -1 +1 @@
-png http://mirrors.xbmc.org/build-deps/sources/libpng-1.5.13.tar.gz
+png http://mirrors.xbmc.org/build-deps/sources/libpng-1.6.26.tar.gz
```
Comment (Rechi) [inline depends/common/png/png.txt:1]:
```suggestion
png http://mirrors.kodi.tv/build-deps/sources/libpng-1.6.26.tar.gz
```

File: lib/Implicit/impSurface.h [hunk 16]
```diff
@@ -0,0 +1,93 @@
...
+#define IMPSURFACE_H
+
+#include <functional>
+
+#ifdef WIN32
+#include <windows.h>
+#endif
+#include <vector>
+
+#include <kodi/gui/gl/GL.h>
```
Comment (Rechi) [inline lib/Implicit/impSurface.h:31]: Why does implicit lib include Kodi headers?

Comment (AlwinEsch) [reply lib/Implicit/impSurface.h:31]: You mean since they are deposited in the addon?

If so, they are temporarily stored in addon and will be added to Kodi M * and removed here.

Comment (Rechi) [reply lib/Implicit/impSurface.h:31]: No my question is not why `kodi/gui/gl/GL.h` and all the other files in lib/kodi are checked into this repo.
My question is, why does a library that doesn't know anything about Kodi, include a Kodi header file?

Comment (AlwinEsch) [reply lib/Implicit/impSurface.h:31]: On the one hand, a lot of changes were needed to make GL4 compatible. The other is, this was designed purely for GL under Windows and therefore had no GLES support. By using the Kodi header, the required one is inserted.

This library can be seen as a death part of the original, so I'm not ashamed to make this suitable for us.
https://github.com/reallyslickscreensavers/rslibs/tree/master/Implicit

File: depends/common/glm/glm.txt (hunk 11 above)
Comment (AlwinEsch) [reply depends/common/glm/glm.txt:1]: Has wrongly looked on https://github.com/g-truc/glm on branch names (highest 0.9.8 and not compatible) and not looked on release (there Version 0.9.9.3, which works) becomes changed.

File: depends/common/gli/gli.txt (hunk 10 above)
Comment (AlwinEsch) [reply depends/common/gli/gli.txt:1]: About `GLI` there is the last release (~2016) no more compatible with current GLM, that's why we come not around to use it, Also GLM noted it somewhere on his Website (not sure where, a bit ago where I read it)

File: CMakeLists.txt (hunk 6 above)
Comment (AlwinEsch) [reply CMakeLists.txt:40]: This was thought as to make a quik linux build with only a `cmake ..` for test ImageMagick and handle his configure command, was easier as over the depends build and thought to leave as fallback if someone compile it itself.

But everything related ImageImagick becomes removed, after next request no more needed.

File: debian/compat (hunk 8 above)
Comment (AlwinEsch) [reply debian/compat:1]: Becomes changed

File: debian/changelog (hunk 7 above)
Comment (AlwinEsch) [reply debian/changelog:1]: Was wrongly added to commit, becomes removed.

File: debian/rules (hunk 9 above)
Comment (AlwinEsch) [reply debian/rules:15]: The standard *.install with:
```
usr/lib/*
usr/share/*
```
does not work on this addon with several addons in one and bring on every addon there the complete content of all (e.g. kodi-screensaver-rsxs-biof), see content: https://pastebin.com/98GQeSpH on a at 23 packages with 77 MB each.

The `*.install.in` with e.g.:
```
usr/lib/@DEB_HOST_MULTIARCH@/kodi/addons/screensaver.rsxs.helios/* /usr/lib/@DEB_HOST_MULTIARCH@/kodi/addons/screensaver.rsxs.helios
usr/share/kodi/addons/screensaver.rsxs.helios/*
```
prevent this and include only the required parts.
